cpp/src/knn/knn_opg_common.cuh

```diff
@@ -68,69 +68,217 @@ namespace opg {
-
-template <typename T>
-void perform_local_operation(T *out, int64_t *knn_indices, T *labels,
-                             size_t cur_batch_size, int k, int n_outputs,
-                             raft::handle_t &h, bool probas_only = false,
-                             std::vector<float *> *probas = nullptr,
-                             std::vector<int *> *uniq_labels = nullptr,
-                             std::vector<int> *n_unique = nullptr);
-
-template <typename T>
-void copy_outputs(T *out, int64_t *knn_indices,
-                  std::vector<std::vector<T *>> &y, size_t cur_batch_size,
-                  int k, int n_outputs, int n_features, int my_rank,
-                  std::vector<Matrix::RankSizePair *> &idxPartsToRanks,
-                  std::shared_ptr<deviceAllocator> alloc, cudaStream_t stream);
+/**
+ * The enumeration of KNN distributed operations
+ */
+enum knn_operation {
+  knn,            /**< Simple KNN */
+  classification, /**< KNN classification */
+  class_proba,    /**< KNN classification probabilities */
+  regression      /**< KNN regression */
+};
+
+/**
+ * A structure to store parameters for distributed KNN
+ */
+struct opg_knn_param {
+  union labels_data {
+    std::vector<std::vector<int *>> *i;
+    std::vector<std::vector<float *>> *f;
+  };
+
+  union outputs_data {
+    std::vector<Matrix::Data<int> *> *i;
+    std::vector<Matrix::Data<float> *> *f;
+  };
+
+  knn_operation knn_op; /**< Type of KNN distributed operation */
+  std::vector<Matrix::Data<int64_t> *> *out_I =
+    nullptr; /**< KNN indices output array */
+  std::vector<Matrix::floatData_t *> *out_D =
+    nullptr; /**< KNN distances output array */
+  std::vector<Matrix::floatData_t *> *idx_data =
+    nullptr; /**< Index input array */
+  Matrix::PartDescriptor *idx_desc =
+    nullptr; /**< Descriptor for index input array */
+  std::vector<Matrix::floatData_t *> *query_data =
+    nullptr; /**< Query input array */
+  Matrix::PartDescriptor *query_desc =
+    nullptr;             /**< Descriptor for query input array */
+  bool rowMajorIndex;    /**< Is index row major? */
+  bool rowMajorQuery;    /**< Is query row major? */
+  size_t k = 0;          /**< Number of nearest neighbors */
+  size_t batch_size = 0; /**< Batch size */
+  bool verbose;          /**< verbose */
+
+  labels_data y;     /**< Labels input array (cl&re) */
+  int n_outputs = 0; /**< Number of outputs per query (cl&re) */
+  outputs_data out;  /**< KNN outputs output array (cl&re) */
+
+  std::vector<int *> *uniq_labels =
+    nullptr; /**< Unique labels (classification) */
+  std::vector<int> *n_unique =
+    nullptr; /**< Number of unique labels (classification) */
+
+  std::vector<std::vector<float *>> *probas =
+    nullptr; /**< KNN classification probabilities output array (class-probas) */
+};
+
+/**
+ * A structure to store utilities for CUDA and RAFT comms
+ */
+struct cuda_utils {
+  cuda_utils(raft::handle_t &handle) {
```

Comment:
Yes, it should be possible to do everything with a raft handle. However it appears that using this structure greatly increase the readability of the code for virtually zero overhead. I can still use raft handle if you think that it's the optimal solution however (maybe for coherence with the rest of the project).

Reply:
It would be nice to see the extra params consolidated and just use the raft handle. I also think about maintainability- and this really is just a duplicate of the raft handle. Also, the raft handle has a method `get_internal_streams()` that should replace the need to copy the internal streams into an additional vector (I built the MNMG design this way so that it can assume the number of internal streams to use is equal to the number available in the handle). We can do this in a future pr though.